[dynamo][guards] Add type info of the guarded value in guard managers

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158805
* #158803
* __->__ #158765

tlparse looks like this

<img width="1165" height="226" alt="image" src="https://github.com/user-attachments/assets/04c4e6b1-34a3-4d9d-8304-6eb6d9a94980" />

This will aid in reading guards.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

anijain2305 committed:
[dynamo][guards] Add type info of the guarded value in guard managers

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Add type info of the guarded value in guard managers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Add type info of the guarded value in guard managers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

Lucaskabela commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -2576,10 +2596,45 @@ class GuardManager {
+        _is_empty_dict(is_empty_dict),
+        _is_immutable(is_immutable),
+        _is_nn_module(is_nn_module),
+        _type_str(type_str) {}
```
Suggest also using std::move for the type_str
```suggestion
        _type_str(std::move(type_str)) {}
```

Lucaskabela commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -3391,8 +3457,17 @@ class DictGuardManager : public GuardManager {
+            is_empty_dict,
+            false, // _is_nn_module
+            false, // _is_immutable
+            type_of),
```
```suggestion
            std::move(type_of)),
```

Lucaskabela reviewed (approved):
Generally LGTM, minor suggestion for using std::move for type_of too

anijain2305 committed:
Update on "[dynamo][guards] Add type info of the guarded value in guard managers"


tlparse looks like this

<img width="1165" height="226" alt="image" src="https://github.com/user-attachments/assets/04c4e6b1-34a3-4d9d-8304-6eb6d9a94980" />

This will aid in reading guards.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge -i
